### PR TITLE
Fix IV button disappearing and GoIV crashing on stop

### LINKED ISSUE #135: Gastly crashes GoIV
1. Start service
2. Try to scan Gastly, 4 out of 5 times the IV button just disappears.
3. Go back to GoIV screen, click STOP button, the application closes with "Unfortunately, GoIV has stopped.".

Samsung Note 5, 6.0.1, GoIV 2.1.0. It didn't happen with the last version from the Play Store, it would just fail to recognize the CP when the smoke covered it.

### PR DESCRIPTION
**Pull request by Blaisorblade**
This PR fixes #135. The various bugs are described in commit messages. There are a few refactorings at the beginning which I did while reviewing possibly relevant code.

### CODE REVIEW

**Comment by Blaisorblade** [comment]
Tested this on top of 33d9f66584248377d4ba5ab2fb4e49e15466ef96, rebased on new master, should still work well. Retesting.
I would also appreciate some reviewing work—this should prevent #135 but I might easily have missed some trickier corner case.

**Comment by Blaisorblade** [comment]
Caveat: if when one presses the IV button we fail to scan a pokemon, now we just show again the IV button. That's what the code tried to do, only now it works, but we should probably tell the user. I'll try informing the user with a Toast...

To reproduce this in testing, I follow @rpgrca's tip:

> By the way, if you are in a Pokemon screen, scroll up the page (so that the CP goes out of scope) and press the IV button, it disappears and when you stop the service it crashes as it did with the Gastly problem.

**Comment by jehy** [comment]
Looks good on first glance.

**Comment by sarav** [comment]
Does this handle battery saver mode? You shouldn't show the IV button in battery saver mode.

**Comment by Blaisorblade** [comment]
> Does this handle battery saver mode? You shouldn't show the IV button in battery saver mode.

Haven't tested that, but after reading the code I think the button shouldn't be shown there and it's hard to mess that. Of course that should still be tested.

**Comment by Blaisorblade** [comment]
@DJCrapsody linked to this report, haven't tested this PR with it yet (will do later). Looking at the code, I think this bug is orthogonal to the changes here.

https://www.reddit.com/r/GoIV/comments/4zarig/bug_if_you_click_the_iv_button_while_the_pokemon/d6urlno

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -215,5 +215,5 @@
     <string name="auto_update_setting_summary">Automatically check for updates when the app starts up</string>
     <string name="check_for_update_setting_title">Check for update</string>
     <string name="check_for_update_setting_summary">See if an update is available</string>
-    <string name="scan_pokemon_failed">Scanning Pokémon failed, scroll to the top and try again!</string>
+    <string name="scan_pokemon_failed">Scanning failed, please scroll up and try again!</string>
```
Comment (Blaisorblade) [inline app/src/main/res/values/strings.xml]: @nahojjjen That's OK with you? I can put what you prefer :-)

Comment (nahojjjen) [inline app/src/main/res/values/strings.xml]: Looks better

**Comment by Blaisorblade** [comment]
Just tested this also in battery saver mode, it seems fine and doesn't show the IV button. The only slight annoyance is that taking a screenshot after scrolling shows no error; however, after that, taking a correct screenshot worked again.

Since otherwise there's consensus, I think we can finally merge this!

**Comment by sarav** [comment]
Btw, a lot of this wouldn't have been needed if we didn't use as many unnecessary Intents I refer to in #162

**Pull request merged**